File: cli/lib/kb/ollama.ex
defmodule Kb.Ollama do
  @moduledoc """
  Minimal Ollama embeddings client using `:httpc` from OTP stdlib (no new deps).

  Assumes Ollama is reachable at `http://localhost:11434`. All failures are
  returned as `{:error, reason}` so the caller can degrade gracefully.
  """

  @default_url "http://localhost:11434/api/embeddings"
  @default_model "nomic-embed-text"
  @timeout_ms 5_000

  @doc """
  Quick reachability check against an Ollama endpoint. Returns `:ok` or `{:error, reason}`.
  Uses the embeddings URL's origin so we can reuse the same config everywhere.
  """
  def ping(url \\ @default_url) do
    origin = origin_of(url)

    :inets.start()

request = {String.to_charlist(origin <> "/api/tags"), []}

    case :httpc.request(:get, request, [{:timeout, 1_000}, {:connect_timeout, 500}], []) do
      {:ok, {{_, status, _}, _headers, _body}} when status in 200..499 -> :ok
      {:ok, _} -> {:error, :bad_status}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Produce an embedding for `text`. Options:

    * `:url`   — embeddings endpoint (default `#{@default_url}`)
    * `:model` — model name (default `#{@default_model}`)
  """
  def embed(text, opts \\ []) when is_binary(text) do
    url = Keyword.get(opts, :url, @default_url)
    model = Keyword.get(opts, :model, @default_model)

    body = Jason.encode!(%{model: model, prompt: text})

    :inets.start()

request =
      {String.to_charlist(url), [{~c"content-type", ~c"application/json"}], ~c"application/json",
       body}

    case :httpc.request(
           :post,
           request,
           [{:timeout, @timeout_ms}, {:connect_timeout, 1_000}],
           []
         ) do
      {:ok, {{_, 200, _}, _headers, resp}} ->
        case Jason.decode(IO.iodata_to_binary(resp)) do
          {:ok, %{"embedding" => vec}} when is_list(vec) -> {:ok, vec}
          {:ok, other} -> {:error, {:unexpected_shape, other}}
          {:error, reason} -> {:error, {:decode, reason}}
        end

      {:ok, {{_, status, _}, _headers, body}} ->
        {:error, {:http, status, IO.iodata_to_binary(body)}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp origin_of(url) do
    uri = URI.parse(url)
    scheme = uri.scheme || "http"
    port = if uri.port, do: ":#{uri.port}", else: ""
    "#{scheme}://#{uri.host || "localhost"}#{port}"
  end
end

File: packages/llm/src/providers/ollama.ts
import type {
  CompleteOptions,
  EmbedOptions,
  EmbeddingProvider,
  LLMProvider,
} from '../types.ts';

const DEFAULT_BASE = 'http://localhost:11434';
const DEFAULT_LLM_MODEL = 'llama3.1';
const DEFAULT_EMBED_MODEL = 'nomic-embed-text';

export interface OllamaConfig {
  baseURL?: string;
  model?: string;
  fetch?: typeof fetch;
}

interface OllamaGenerateResponse {
  response: string;
}

interface OllamaEmbedResponse {
  embedding: number[];
}

export function ollamaLLM(config: OllamaConfig = {}): LLMProvider {
  const baseURL = config.baseURL ?? DEFAULT_BASE;
  const defaultModel = config.model ?? DEFAULT_LLM_MODEL;
  const fetchImpl = config.fetch ?? fetch;

  return {
    name: 'ollama',
    async complete(prompt: string, opts: CompleteOptions = {}): Promise<string> {
      const res = await fetchImpl(`${baseURL}/api/generate`, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({
          model: opts.model ?? defaultModel,
          prompt,
          system: opts.system,
          stream: false,
          options: { num_predict: opts.maxTokens ?? 1024 },
        }),
      });
      if (!res.ok) {
        throw new Error(`ollama: HTTP ${res.status}: ${await res.text()}`);
      }
      const json = (await res.json()) as OllamaGenerateResponse;
      return json.response;
    },
  };
}

export function ollamaEmbeddings(config: OllamaConfig = {}): EmbeddingProvider {
  const baseURL = config.baseURL ?? DEFAULT_BASE;
  const defaultModel = config.model ?? DEFAULT_EMBED_MODEL;
  const fetchImpl = config.fetch ?? fetch;

  async function one(text: string, opts: EmbedOptions = {}): Promise<Float32Array> {
    const res = await fetchImpl(`${baseURL}/api/embeddings`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ model: opts.model ?? defaultModel, prompt: text }),
    });
    if (!res.ok) {
      throw new Error(`ollama-embeddings: HTTP ${res.status}: ${await res.text()}`);
    }
    const json = (await res.json()) as OllamaEmbedResponse;
    return new Float32Array(json.embedding);
  }

  return {
    name: 'ollama-embeddings',
    async embed(text, opts) {
      return one(text, opts);
    },
    async embedBatch(texts, opts) {
      const out: Float32Array[] = [];
      for (const t of texts) out.push(await one(t, opts));
      return out;
    },
  };
}
